### PR TITLE
ad7768-1fmcz: ad77681: added macros and functions on ad7761 noos library

### PR DESCRIPTION
**Pull request by jgarlitos**
Signed-off-by: JohnNeeko Garlitos <JohnNeeko.Garlitos@analog.com>

### CODE REVIEW

File: ad7768-1fmcz/ad77681.c [hunk 1]
```diff
@@ -43,6 +43,7 @@
 #include "stdio.h"
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:43]: A more elaborate commit comment, should be added

File: ad7768-1fmcz/ad77681.c [hunk 2]
```diff
@@ -118,9 +124,24 @@ int32_t ad77681_spi_reg_read(struct ad77681_dev *dev,
...
+		crc_buf[1] = buf[1];
+		
+		if (dev->crc_sel == AD77681_XOR)
+			crc = ad77681_compute_xor(crc_buf, 2, INITIAL_CRC);								// INITIAL_CRC is 0, wen ADC is not in continuous-read mode
+		
+		else if(dev->crc_sel == AD77681_CRC)
+			crc = ad77681_compute_crc8(crc_buf, 2, INITIAL_CRC);							// INITIAL_CRC is 0, wen ADC is not in continuous-read mode
+		
+		if (crc != buf[2])																	// In buf[2] is CRC from the ADC
+			ret = -1;
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:139]: Return `FAILURE`, from `error.h`

File: ad7768-1fmcz/ad77681.c [hunk 3]
```diff
@@ -134,12 +155,16 @@ int32_t ad77681_spi_reg_write(struct ad77681_dev *dev,
 			      uint8_t reg_addr,
 			      uint8_t reg_data)
 {
-	uint8_t buf[2];
+	uint8_t buf[3];
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:158]: This change has no business, with the changes in `ad77681_compute_crc8()`function.
This commit, should be split in multiple commits, by functionalities that it touches.

File: ad7768-1fmcz/ad77681.c [hunk 4]
```diff
@@ -118,9 +124,24 @@ int32_t ad77681_spi_reg_read(struct ad77681_dev *dev,
 	if (ret < 0)
 		return ret;
 
+	if (dev->crc_sel != AD77681_NO_CRC)														// XOR or CRC checksum for read transactions
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:127]: Comments should be like this: `/*  XOR or CRC checksum for read transactions */`

File: ad7768-1fmcz/ad77681.c [hunk 5]
```diff
@@ -118,9 +124,24 @@ int32_t ad77681_spi_reg_read(struct ad77681_dev *dev,
...
+	{
+		crc_buf[0] = AD77681_REG_READ(reg_addr);
+		crc_buf[1] = buf[1];
+		
+		if (dev->crc_sel == AD77681_XOR)
+			crc = ad77681_compute_xor(crc_buf, 2, INITIAL_CRC);								// INITIAL_CRC is 0, wen ADC is not in continuous-read mode
+
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:134]: I would remove this line between `if` statement and `else` statement.

File: ad7768-1fmcz/ad77681.c [hunk 6]
```diff
@@ -243,10 +313,321 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
 
 	/* Fill the adc_data buffer */
 	memcpy(adc_data, buf, ARRAY_SIZE(buf));
+	
+	if (buf[1] & 0x80)
+		converted_data = (int32_t)(0xFF << 24 | (((buf[1]  &   0xFF)  << 16) |
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:318]: add brackets for `(0xFF << 24)` operation. Same as you did for `((buf[1]  &   0xFF)  << 16)`

File: ad7768-1fmcz/ad77681.c [hunk 7]
```diff
@@ -211,30 +235,76 @@ uint8_t ad77681_get_rx_buf_len(struct ad77681_dev *dev)
 }
 
 /**
+ * Helper function to get the number of SPI bytes for INTERRUPT ADC DATA READ
+ * @param dev - The device structure.
+ * @return frame_bytes - the number of rx bytes
+ */
+uint8_t ad77681_get_frame_bytes(struct ad77681_dev *dev)
+{
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:243]: This implementation, should be in a separate commit.

File: ad7768-1fmcz/ad77681.c [hunk 8]
```diff
@@ -243,10 +313,321 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+
+	return ret;
+}
+
+/**
+ * This function uses continuous read mode, which has to be enabled in advance
+ * Read conversion result from device using GPIO DRDY interrupt in 16bit frames
+ *
+ *	    /-------------\       /-------------\       /-------------\
+ *------  24-bit data  -------  8-bit staus  -------   8-bit CRC   -------
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:340]: I would avoid creating diagrams from text characters for the header function comments. Here looks fine, but when Doxygen will parse the text, it will be messed up.

File: ad7768-1fmcz/ad77681.c [hunk 9]
```diff
@@ -243,10 +313,321 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+	
+	buf[0] = 0;																					// 16bit buffers! Used only for ADC data read
+	buf[1] = 0;
+	buf[2] = 0;
+	
+	ret = spi_read_data(dev->spi_desc, buf, dev->data_frame_bytes);
+	if (ret < 0)
+		return ret;
+		
+	if (dev->conv_len == AD77681_CONV_24BIT)													// 24bit format
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:367]: please add comments like `// Cuts first 8 bits from buf[1] and paste it after 16bits of buf[0]` above the code that handles that specific comment. The number of characters per line should not be above 80 (only in exceptional situations).

check for similar issues throughout the file.

File: ad7768-1fmcz/ad77681.c [hunk 10]
```diff
@@ -211,30 +235,76 @@ uint8_t ad77681_get_rx_buf_len(struct ad77681_dev *dev)
 }
 
 /**
+ * Helper function to get the number of SPI bytes for INTERRUPT ADC DATA READ
+ * @param dev - The device structure.
+ * @return frame_bytes - the number of rx bytes
+ */
+uint8_t ad77681_get_frame_bytes(struct ad77681_dev *dev)
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:242]: I think this function is used only local. So, probably it should be removed from header, and make it static.

File: ad7768-1fmcz/ad77681.c [hunk 11]
```diff
@@ -243,10 +313,321 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad77681_data_to_voltage(struct ad77681_dev *dev,
+	struct adc_data *measured_data,
+	double *voltage)
+{
+	int32_t converted_data = 0;
+	
+	for (measured_data->count = 0; measured_data->count < measured_data->samples; measured_data->count++)
+	{
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:478]: as mentioned above for comments, try to fit the code into a maximum of 80 characters per line.

check for similar issues throughout the entire code.

File: ad7768-1fmcz/ad77681.c [hunk 12]
```diff
@@ -282,13 +665,291 @@ int32_t ad77681_set_power_mode(struct ad77681_dev *dev,
...
+ * @param dev   -  The device structure.
+ * @param enable - The positive reference buffer selector
+ * 					Accepted values: true
+ *									 false
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad77681_set_50HZ_rejection(struct ad77681_dev *dev,
+	uint8_t enable)
+{
+	int32_t ret;
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:891]: maybe directly `return ad77681_spi_write_mask`.

No need for extra variable `ret` in this situation.

File: ad7768-1fmcz/ad77681.c [hunk 13]
```diff
@@ -211,30 +235,76 @@ uint8_t ad77681_get_rx_buf_len(struct ad77681_dev *dev)
...
  * Read conversion result from device.
  * @param dev - The device structure.
  * @param adc_data - The conversion result data
+ * @param voltage  - Measured data transfered to voltage
  * @return 0 in case of success, negative error code otherwise.
  */
 int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
-				  uint8_t *adc_data)
+				uint8_t *adc_data,
+				double *voltage)
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:277]: This breaks the interface, other projects, that uses this function, wont compile(Ex:` ad77681_evb.c`)
A separate commit should be added with changes in `ad77681_evb.c`

File: ad7768-1fmcz/ad77681.c [hunk 14]
```diff
@@ -446,23 +1109,761 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+			AD77681_REG_SYNC_RESET,
+			AD77681_SYNC_RST_SPI_RESET_MSK,
+			AD77681_SYNC_RST_SPI_RESET(0x3));
+
+		ret |= 	ad77681_spi_write_mask(dev,
+			AD77681_REG_SYNC_RESET,
+			AD77681_SYNC_RST_SPI_RESET_MSK,
+			AD77681_SYNC_RST_SPI_RESET(0x2));		
+	}	
+
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:1136]: remove empty line between `if` and `else`

File: ad7768-1fmcz/ad77681.c [hunk 15]
```diff
@@ -446,23 +1109,761 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad77681_reset(struct ad77681_dev *dev,
+	enum ad7761_reset_option rst_option)
+{
+	int32_t ret = 0;
+
+	if (rst_option == AD77681_SOFT_RESET)							// Soft reset over SPI
+	{		
+		ret |= 	ad77681_spi_write_mask(dev,							// Two writes are required to initialize the reset
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:1126]: Here you should initialize directly the ret variable, without performing the `or` operation.

File: ad7768-1fmcz/ad77681.c [hunk 16]
```diff
@@ -243,10 +313,321 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+		mclk_div = 4;
+		break;
+	case AD77681_MCLK_DIV_2:
+		mclk_div = 2;
+		break;
+	default:
+		return -1;
+		break;
+	}
+
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:520]: nitpick: extra line

File: ad7768-1fmcz/ad77681.c [hunk 17]
```diff
@@ -243,10 +313,321 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+		mclk_div = 2;
+		break;
+	default:
+		return -1;
+		break;
+	}
+	
+	
+	// Finding out decimation ratio
+	switch (dev->filter)
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:523]: similar to the `switch (dev->decimate)` statement, you should not have empty lines between cases.

File: ad7768-1fmcz/ad77681.c [hunk 18]
```diff
@@ -243,10 +313,321 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+	return ret;			
+}
+
+/**
+ * CRC and status bit handling after each readout form the ADC
+ * @param dev - The device structure.
+ * @param *data_buffer - 16-bit buffer readed from the ADC containing the CRC, data and the stattus bit
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad77681_CRC_status_handling(struct ad77681_dev *dev,
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:385]: Is this function used outside of this file? If not make it static and remove it from header.

File: ad7768-1fmcz/ad77681.c [hunk 19]
```diff
@@ -282,13 +665,291 @@ int32_t ad77681_set_power_mode(struct ad77681_dev *dev,
...
+	enum ad77681_REFn_buffer REFn)
+{
+	int32_t ret;
+	
+	ret = ad77681_spi_write_mask(dev,
+		AD77681_REG_ANALOG,
+		AD77681_ANALOG_REF_BUF_NEG_MSK,
+		AD77681_ANALOG_REF_BUF_NEG(REFn));
+
+	dev->REFn = REFn;
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:774]: move this line of code above `ad77681_spi_write_mask` and return directly the function. (no need for extra variable `ret`).

File: ad7768-1fmcz/ad77681.c [hunk 20]
```diff
@@ -282,13 +665,291 @@ int32_t ad77681_set_power_mode(struct ad77681_dev *dev,
...
+	enum ad77681_REFp_buffer REFp)
+{
+	int32_t ret;
+	
+	ret = ad77681_spi_write_mask(dev,
+		AD77681_REG_ANALOG,
+		AD77681_ANALOG_REF_BUF_POS_MSK,
+		AD77681_ANALOG_REF_BUF_POS(REFp));
+
+	dev->REFp = REFp;
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:798]: move this line of code above ad77681_spi_write_mask and return directly the function. (no need for extra variable ret).

File: ad7768-1fmcz/ad77681.c [hunk 21]
```diff
@@ -282,13 +665,291 @@ int32_t ad77681_set_power_mode(struct ad77681_dev *dev,
...
+			AD77681_REG_DIGITAL_FILTER,
+			AD77681_DIGI_FILTER_FILTER_MSK,
+			AD77681_DIGI_FILTER_FILTER(filter));
+		
+		ret |= ad77681_spi_write_mask(dev,
+			AD77681_REG_DIGITAL_FILTER,
+			AD77681_DIGI_FILTER_DEC_RATE_MSK,
+			AD77681_DIGI_FILTER_DEC_RATE(decimate));
+	}
+
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:851]: nitpick: extra line

File: ad7768-1fmcz/ad77681.c [hunk 22]
```diff
@@ -243,10 +313,321 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+					checksum_byte = data_buffer[2] >> 8;
+					checksum_length = 4;
+				}		
+			else																				// 16bit ADC data + 8bit of status + 8bit of CRC = 2 16bit frames
+				{
+					checksum_byte = data_buffer[1] & 0xFF;
+					checksum_length = 3;
+				}	
+		}
+
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:417]: Extra line here.

File: ad7768-1fmcz/ad77681.c [hunk 23]
```diff
@@ -446,23 +1109,761 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+/**
+ * Initiate a SYNC_OUT pulse over spi
+ * @param dev 			The device structure.
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad77681_initiate_sync(struct ad77681_dev *dev)
+{
+	int32_t ret;
+
+	ret = ad77681_spi_write_mask(dev,
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:1156]: return directly `ad77681_spi_write_mask`

File: ad7768-1fmcz/ad77681.c [hunk 24]
```diff
@@ -243,10 +313,321 @@ int32_t ad77681_spi_read_adc_data(struct ad77681_dev *dev,
...
+				}		
+			else																				// 16bit ADC data + 8bit of status + 8bit of CRC = 2 16bit frames
+				{
+					checksum_byte = data_buffer[1] & 0xFF;
+					checksum_length = 3;
+				}	
+		}
+		
+		else																					// without status bit
+			{
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:419]: Alignment seems to be wrong.

File: ad7768-1fmcz/ad77681.c [hunk 25]
```diff
@@ -446,23 +1109,761 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+	uint16_t twait;
+	int32_t twos_complement, ret = 0;
+	
+	const uint8_t coeff_reg_length = 56;
+	const uint8_t key1 = 0xAC, key2 = 0x45, key3 = 0x55;						// Specifice keys taken from the AD7768-1 datasheet as a part of the upload sequence
+	const float coeff_scale_factor = (1 << 22);									// Scaling factor for all coefficients 2^22
+	
+	twait = (uint16_t)(((512.0) / ((float)(dev->mclk))) * 1000.0) + 1; 			// Wait time in uS necessary for access the COEFF_CONTROL and COEFF_DATA registers
+																				// Equation taken form the AD7768-1 datasheet => Twait = 512/MCLK
+
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:1258]: nitpick: extra line

File: ad7768-1fmcz/ad77681.c [hunk 26]
```diff
@@ -446,23 +1109,761 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+int32_t ad77681_gpio_read(struct ad77681_dev *dev,
+	uint8_t *value,
+	enum ad77681_gpios gpio_number)
+{
+	int32_t ret = 0;
+	
+	switch (gpio_number)
+	{
+	case AD77681_GPIO0:
+
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:1387]: no empty line between cases.

File: ad7768-1fmcz/ad77681.c [hunk 27]
```diff
@@ -446,23 +1109,761 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+int32_t ad77681_gpio_inout(struct ad77681_dev *dev,
+	uint8_t direction,
+	enum ad77681_gpios gpio_number)
+{
+	int32_t ret = 0;
+
+	switch (gpio_number)
+	{			
+	case AD77681_GPIO0:											// Set direction of GPIO0
+		ret |= ad77681_spi_write_mask(dev,
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:1522]: no need for `or` operation. only one case will be accessed per function call.

Comment (amiclaus) [reply ad7768-1fmcz/ad77681.c:1522]: Same for all cases.

File: ad7768-1fmcz/ad77681.c [hunk 28]
```diff
@@ -446,23 +1109,761 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+ *						Accepted values: AD77681_GLOBAL_GPIO_ENABLE
+ *										 AD77681_GLOBAL_GPIO_DISABLE
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad77681_global_gpio(struct ad77681_dev *dev,
+	enum ad77681_gobal_gpio_enable gpio_enable)
+{
+	int32_t ret;
+
+	ret = ad77681_spi_write_mask(dev,
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:1578]: return directly `ad77681_spi_write_mask`

File: ad7768-1fmcz/ad77681.c [hunk 29]
```diff
@@ -446,23 +1109,761 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+		AD77681_GPIO_CNTRL_UGPIO_EN_MSK,
+		AD77681_GPIO_CNTRL_UGPIO_EN(gpio_enable));
+
+	return ret;	
+}
+
+/**
+ * Read and write from ADC scratchpad register to check SPI Communication in the very beginning, during inicialization.
+ * @param dev 			The device structure.
+ * @param *sequence		The sequence which will be written into scratchpad and the readed sequence will be returned
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:1589]: no need to specify that the parameter is a pointer in the function comment.

File: ad7768-1fmcz/ad77681.c [hunk 30]
```diff
@@ -446,23 +1109,761 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+	status->spi_clock_count = buf[1] & AD77681_SPI_CLK_CNT_ERROR_MSK;
+	status->spi_read_error = buf[1] & AD77681_SPI_READ_ERROR_MSK;
+	status->spi_write_error = buf[1] & AD77681_SPI_WRITE_ERROR_MSK;
+	status->spi_crc_error = buf[1] & AD77681_SPI_CRC_ERROR_MSK;
+	
+	ret |= ad77681_spi_reg_read(dev, AD77681_REG_ADC_DIAG_STATUS, buf);  					// ADC diag status register
+	
+	status->dldo_psm_error = buf[1] & AD77681_ADC_DLDO_PSM_ERROR_MSK;
+	status->aldo_psm_error = buf[1] & AD77681_ADC_ALDO_PSM_ERROR_MSK;
+    status->ref_det_error = buf[1] & AD77681_ADC_REF_DET_ERROR_MSK;
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:1826]: looks like the code lines are not properly indented here.

File: ad7768-1fmcz/ad77681.c [hunk 31]
```diff
@@ -266,6 +647,8 @@ int32_t ad77681_set_power_mode(struct ad77681_dev *dev,
 				      AD77681_POWER_CLK_PWRMODE_MSK,
 				      AD77681_POWER_CLK_PWRMODE(mode));
 
+	dev->power_mode = mode;
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:650]: This seems like a fix. Add a separate commit for it.
If `ad77681_spi_write_mask` fails, is `dev->power_mode` is still set.

File: ad7768-1fmcz/ad77681.c [hunk 32]
```diff
@@ -282,13 +665,291 @@ int32_t ad77681_set_power_mode(struct ad77681_dev *dev,
 int32_t ad77681_set_mclk_div(struct ad77681_dev *dev,
 			     enum ad77681_mclk_div clk_div)
 {
-	ad77681_spi_write_mask(dev,
+	int32_t ret = 0;
+
+	ret |= ad77681_spi_write_mask(dev,
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:670]: Same as above.

File: ad7768-1fmcz/ad77681.c [hunk 33]
```diff
@@ -282,13 +665,291 @@ int32_t ad77681_set_power_mode(struct ad77681_dev *dev,
...
+ *									 AD77681_VCM_2_5V
+	  								 AD77681_VCM_2_05V
+									 AD77681_VCM_1_9V
+									 AD77681_VCM_1_65V
+									 AD77681_VCM_1_1V
+									 AD77681_VCM_0_9V
+									 AD77681_VCM_OFF
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad77681_set_VCM_output(struct ad77681_dev *dev,
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:694]: Add a separate commit with all new implementations.

File: ad7768-1fmcz/ad77681.c [hunk 34]
```diff
@@ -282,13 +665,291 @@ int32_t ad77681_set_power_mode(struct ad77681_dev *dev,
...
+	enum ad77681_VCM_out VCM_out)
+{
+	int32_t ret;
+	
+	ret = ad77681_spi_write_mask(dev,
+		AD77681_REG_ANALOG2,
+		AD77681_ANALOG2_VCM_MSK,
+		AD77681_ANALOG2_VCM(VCM_out));
+
+	dev->VCM_out = VCM_out;
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:704]: Check return value, before doing anything else.
Check in code for similar issues.

File: ad7768-1fmcz/ad77681.c [hunk 35]
```diff
@@ -282,13 +665,291 @@ int32_t ad77681_set_power_mode(struct ad77681_dev *dev,
...
+ *									 AD77681_BUFn_ENABLED
+ *									 AD77681_BUFn_FULL_BUFFER_ON
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad77681_set_continuos_read(struct ad77681_dev *dev,
+	enum ad77681_continuous_read continuous_enable)
+{
+	int32_t ret;
+	
+	if (continuous_enable)
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:915]: Add brace if else statement has.

File: ad7768-1fmcz/ad77681.c [hunk 36]
```diff
@@ -315,17 +976,19 @@ int32_t ad77681_set_conv_mode(struct ad77681_dev *dev,
 			      enum ad77681_conv_diag_mux diag_mux_sel,
 			      bool conv_diag_sel)
 {
-	ad77681_spi_write_mask(dev,
+	int32_t ret;
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:979]: Add separate commit for each fix.

File: ad7768-1fmcz/ad77681.c [hunk 37]
```diff
@@ -282,13 +665,291 @@ int32_t ad77681_set_power_mode(struct ad77681_dev *dev,
...
+ */
+
+int32_t ad77681_set_filter_type(struct ad77681_dev *dev,
+	enum ad77681_sinc5_fir_decimate decimate,
+	enum ad77681_filter_type filter,
+	uint16_t sinc3_osr)
+{
+	int32_t ret = 0;
+	
+	ret |= ad77681_spi_reg_write(dev, AD77681_REG_DIGITAL_FILTER, 0x00);
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:832]: no need for `or` operation here.

File: ad7768-1fmcz/ad77681.c [hunk 38]
```diff
@@ -446,23 +1109,761 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+	{
+		// Set the coeff address, write 0xC0 to COEF_CONTROLL
+		ret |= ad77681_spi_reg_write(dev,
+			AD77681_REG_COEFF_CONTROL,
+			address);
+		
+		// 4. the address is set, now wait 
+		_delay_us(twait);		
+																				// The COEFF_DATA register has to be filled with 56 coeffs!
+																				// In case of usage less than 56 active coeffs, the remaining
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:1301]: nitpick: extra line

File: ad7768-1fmcz/ad77681.c [hunk 39]
```diff
@@ -446,23 +1109,761 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+										
+			twos_complement = (int32_t)(coeffs[i] * coeff_scale_factor);		// Scaling value of each coefficient * 2^22		
+			
+			// 5. wirte coefficients to COEFF_DATA
+			coeffs_buf[0] = AD77681_REG_WRITE(AD77681_REG_COEFF_DATA);
+			coeffs_buf[1] = (twos_complement & 0xFF0000) >> 16;
+			coeffs_buf[2] = (twos_complement & 0x00FF00) >> 8;
+			coeffs_buf[3] = (twos_complement & 0x0000FF);
+		}
+
```
Comment (amiclaus) [inline ad7768-1fmcz/ad77681.c:1313]: remove extra line between if and else conditions

File: ad7768-1fmcz/ad77681.c [hunk 40]
```diff
@@ -446,23 +1109,761 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+	
+	ret |= ad77681_spi_write_mask(dev,
+		AD77681_REG_GAIN_LO,
+		AD77681_GAIN_LOW_MSK,
+		AD77681_GAIN_LOW(gain_LO));
+	
+	return ret;
+}
+
+
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:1235]: Extra lines

File: ad7768-1fmcz/ad77681.c [hunk 41]
```diff
@@ -446,23 +1109,761 @@ int32_t ad77681_soft_reset(struct ad77681_dev *dev)
...
+/**
+ * Programmamble fir filter
+ * @param dev 				The device structure.
+ * @param *coefs			Desired coefficients in array of the filter to be written
+ * @param count_of_coeffs	Count of active coeffs
+ * @return 0 in case of success, negative error code otherwise.
+ */
+int32_t ad77681_programmable_filter(struct ad77681_dev *dev,
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:1244]: Maybe break this functions into smaller ones.

File: ad7768-1fmcz/ad77681.c [hunk 42]
```diff
@@ -471,24 +1872,74 @@ int32_t ad77681_setup(struct ad77681_dev **device,
...
+	dev->REFp = init_param.REFp;	
+	dev->filter = init_param.filter;
+	dev->decimate = init_param.decimate;
+	dev->sinc3_osr = init_param.sinc3_osr;
+	dev->vref = init_param.vref;
+	dev->mclk = init_param.mclk;
+	dev->sample_rate = init_param.sample_rate;
+	dev->data_frame_bytes = init_param.data_frame_bytes;
+
+	spi_init(&dev->spi_desc, &init_param.spi_eng_dev_init);
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:1888]: Check the return value.

File: ad7768-1fmcz/ad77681.c [hunk 43]
```diff
@@ -471,24 +1872,74 @@ int32_t ad77681_setup(struct ad77681_dev **device,
...
+		ad77681_update_sample_rate(dev);
+		*device = dev;
+	}
 	if (!ret)
 		printf("ad77681 successfully initialized\n");
 
 	return ret;
 }
+
+int32_t ad77681_reset_spi_interface(void)
```
Comment (cristiantpop) [inline ad7768-1fmcz/ad77681.c:1937]: Add function comment

**Comment by jgarlitos** [comment]
Hi All,
Thanks for the review. I think it would be better to raised another pull request for this one. I'll separate the commits for the different functionalities.
